// @typeit/discord/decorators/classes/DDiscord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DDiscord = void 0;
const Decorator_1 = require("./Decorator");
const __1 = require("../..");
class DDiscord extends Decorator_1.Decorator {
    constructor() {
        super();
        this._guards = [];
        this._slashes = [];
        this._events = [];
        this._defaultPermission = true;
        this._permissions = [];
        this._guilds = [];
    }
    get permissions() {
        return this._permissions;
    }
    set permissions(value) {
        this._permissions = value;
    }
    get guilds() {
        return this._guilds;
    }
    set guilds(value) {
        this._guilds = value;
    }
    get defaultPermission() {
        return this._defaultPermission;
    }
    set defaultPermission(value) {
        this._defaultPermission = value;
    }
    get description() {
        return this._description;
    }
    set description(value) {
        this._description = value;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get guards() {
        return this._guards;
    }
    set guards(value) {
        this._guards = value;
    }
    get slashes() {
        return this._slashes;
    }
    set slashes(value) {
        this._slashes = value;
    }
    get events() {
        return this._events;
    }
    set events(value) {
        this._events = value;
    }
    get instance() {
        return __1.DIService.instance.getService(this.from);
    }
    static create(name) {
        const discord = new DDiscord();
        discord.name = name;
        return discord;
    }
}
exports.DDiscord = DDiscord;
