// @typeit/discord/decorators/classes/Decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Decorator = void 0;
const logic_1 = require("../../logic");
class Decorator {
    constructor() {
        this._index = undefined;
    }
    get index() {
        return this._index;
    }
    get classRef() {
        return this._classRef;
    }
    set classRef(value) {
        this._classRef = value;
        this.from = value;
    }
    get from() {
        return this._from;
    }
    set from(value) {
        this._from = value;
    }
    get key() {
        return this._key;
    }
    get method() {
        return this._method;
    }
    get isClass() {
        return !!this._method;
    }
    decorateUnknown(classRef, key, method, index) {
        const decorateAClass = logic_1.DecoratorUtils.decorateAClass(method) && index === undefined;
        const finalClassRef = decorateAClass
            ? classRef
            : classRef.constructor;
        const finalKey = decorateAClass ? finalClassRef.name : key;
        const finalMethod = decorateAClass ? finalClassRef : method === null || method === void 0 ? void 0 : method.value;
        return this.decorate(finalClassRef, finalKey, finalMethod, finalClassRef, index);
    }
    decorate(classRef, key, method, from, index) {
        this._from = from || classRef;
        this._classRef = classRef;
        this._key = key;
        this._method = method;
        this._index = index;
        this.update();
        return this;
    }
    update() { }
}
exports.Decorator = Decorator;
